### PR TITLE
Add e2e to verify runAsNonRoot option in pod's securityContext

### PR DESCRIPTION
**Pull request by mgdevstack**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug

/kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
Adds an e2e to validate `runAsNonRoot` securityContext option. A conformant platform may not allow to run container with uid 0 (root user) and still be conformant. So added new e2e instead of tweaking existing e2e ```should run the container with uid 65534```.
**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:
New e2e `should not allow to run the container as root when RunAsNonRoot securityContext is set` is created so that we can promote this to Conformance as it is an important point to consider that _"A conformant platform may not allow to run container with uid 0 (root user) and still be conformant"_
Reference: https://github.com/kubernetes/kubernetes/pull/70639#discussion_r257870294

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```
/area conformance
/sig testing
@kubernetes/sig-node-pr-reviews

### CODE REVIEW

**Comment by tallclair** [comment]
/lgtm

It would be good to have a test for runAsNonRoot when the image specifies the user, but I'm not sure whether that should be a conformance test...

File: test/e2e/common/security_context.go [hunk 1]
```diff
@@ -50,44 +53,92 @@ var _ = framework.KubeDescribe("Security Context", func() {
...
+
+			err := wait.Poll(5*time.Second, 2*time.Minute, func() (bool, error) {
+				kubeletPod, err := f.PodClient().Get(podName, metav1.GetOptions{})
+				Expect(err).NotTo(HaveOccurred(), "Pod must exists")
+				for _, status := range kubeletPod.Status.ContainerStatuses {
+					if status.Name == podName {
+						switch {
+						case status.State.Waiting != nil:
+							if status.State.Waiting.Reason == "CreateContainerConfigError" &&
+								strings.Contains(status.State.Waiting.Message, "runAsUser breaks non-root policy") {
```
Comment (spiffxp) [inline test/e2e/common/security_context.go]: [conformance-test-requirements](https://github.com/kubernetes/community/blob/20d3fe4cd54d65f74a13b37d2648ffefa067a4c9/contributors/devel/sig-architecture/conformance-tests.md#conformance-test-requirements) state:
>Examples of tests which are not eligible for promotion to conformance:
>- anything that checks optional Condition fields, such as Reason or Message, as these may change over time (however it is reasonable to verify these fields exist or are non-empty)

**Review by spiffxp — COMMENTED** [review]
This may be a reasonable test, but as written it's not going to be eligible for Conformance, how would you like to proceed?

**Comment by spiffxp** [comment]
/remove-priority important-soon
/priority important-longterm
(taking a guess that this may not land this release)

File: test/e2e/common/security_context.go (hunk 1 above)
Comment (mgdevstack) [reply test/e2e/common/security_context.go]: Though it is using optional option `RunAsNonRoot ` from podSpec but very important for a Conformant platform which may not allow a root user to run container.

As this e2e is meant to go for Conformance, I can update code to check non-empty reason and message. (**Disadvantage:** no accuracy of e2e as covering broader reason of Container's waiting state)

_Other suggestions are highly appreciated._

**Comment by spiffxp** [comment]
/hold
I think this would be worth discussing at the upcoming [conformance office hours meeting](https://github.com/kubernetes/community/tree/master/sig-architecture#meetings)

**Comment by timothysc** [comment]
/lgtm cancel 

Top of review column for next meeting now.

File: test/e2e/common/security_context.go (hunk 1 above)
Comment (timothysc) [reply test/e2e/common/security_context.go]: @bgrant0607 - here is an example of an ask, but given that the tests are only valid for their specific release are there use cases that actually exist where 

> anything that checks optional Condition fields, such as Reason or Message, as these may change over time (however it is reasonable to verify these fields exist or are non-empty) 

actually apply, and do we want to allow that?

Comment (timothysc) [reply test/e2e/common/security_context.go]: /assign @bgrant0607

**Comment by bgrant0607** [comment]
This test is overly specific. What the user cares about is that the container isn't run as root. The message could change. Or the pod could fail, either in validation, admission control, or at the Kubelet, and it would still be acceptable behavior.

**Comment by mgdevstack** [comment]
Thanks for clarity. I would update the codebase accordingly.

**Comment by mgdevstack** [comment]
@spiffxp Can we remove hold now? New code updates has been pushed.(ready for review)

**Comment by spiffxp** [comment]
/approve
no longer looks for a specific message, but instead confirms that the pod actually ran (which it would only do if it runs as non-root)

needs rebase

**Comment by spiffxp** [comment]
/lgtm

Leaving milestone off for now, I'm not yet clear whether we want to push for this to merge during code freeze

**Comment by tallclair** [comment]
This test is redundant with https://github.com/kubernetes/kubernetes/pull/78264 now

**Pull request closed without merging**